zkml/src/layers/transformer/softmax.rs

```diff
@@ -1,60 +1,361 @@
+    /// The proof for the error table lookup
+    pub(crate) error_lookup: LogUpProof<E>,
+    /// Witness commitments for this layer
+    pub(crate) commitments: Vec<PCS::Commitment>,
+    /// The sumcheck proof we use to make sure everything is evaluated at the same point.
+    pub(crate) accumulation_proof: IOPProof<E>,
+    /// Sumcheck proof used to prove that the mask was applied correctly
+    pub(crate) mask_proof: IOPProof<E>,
+    pub(crate) tril_eval: E,
+    /// The claimed evaluations of the commitments
+    pub(crate) evaluations: Vec<E>,
+}
+
+impl<N: Number> Default for Softmax<N> {
+    fn default() -> Self {
+        Softmax {
+            scalar: N::unit(),
+            max_size: 1024usize,
+            quant_info: None,
+        }
+    }
 }
 
 impl<N: Number> Softmax<N> {
     pub fn new() -> Self {
-        Self {
-            scale: N::unit(),
-            apply_on_dim: None,
+        Softmax::<N>::default()
+    }
+
+    pub fn new_with_scale(scale: N, max_context_size: usize) -> Softmax<N> {
+        Softmax {
+            scalar: scale,
+            max_size: max_context_size,
+            quant_info: None,
         }
     }
-    pub fn with_scale(self, scale: N) -> Self {
-        Self { scale, ..self }
+    pub fn quantise(&self, input_scaling: ScalingFactor) -> Result<Softmax<Element>> {
+        // First we work out what we need to multiply by to get the input scale factor to be 2^32
+        let input_scale_factor = input_scaling.scale();
+        let temperature = self.scalar.to_f32()?;
+        let float_temperature = 1.0f32 / temperature;
+        let multiplier = (SCALE_FACTOR as f32 * input_scale_factor).round() as Element;
+
+        // minimum_input is calculated as `(input_min - sqrt(d) * ln_n - d * input_max)/sqrt(d)` and then quantised
+        let input_min = input_scaling.min();
+        let input_max = input_scaling.max();
+
+        let min_input_float = input_min
+            - float_temperature * (self.max_size as f32 * (input_max * temperature).exp()).ln();
+        // Now that we have the minimum possible input as a float we need to work out how many integral bits we need to account for
+        // We know that the minimum input is negative so first we take the absoloute value
+        let min_input_abs = min_input_float.abs();
+
+        let int = min_input_abs.round() as usize;
+        let integral_bits = ceil_log2(int);
+
+        let base = 1i128 << (LOG_SCALE_FACTOR - 8);
+
+        let (float_error, bkm_float) = calc_softmax_error(
+            base,
+            self.max_size as f32,
+            OUTPUT_SCALE_FACTOR as f32,
+            SCALE_FACTOR as f32,
+            3.0f32,
+            0.0f32,
+            2.0f32,
+            float_temperature,
+        );
+
+        let float_error = float_error.abs();
+        let bkm = bkm_float.round() as Element;
+
+        let table_bits = (integral_bits + 8).max(ceil_log2(bkm as usize >> 16));
+        let table_size = 1i128 << table_bits;
```

Comment:
should we make a quick assertion that our table size is not too big ? Given bkm is given by the user, it might go big no ?